utils/small_vector.hh

```diff
@@ -144,6 +145,17 @@ public:
         }
     }
 
+    small_vector(std::from_range_t, std::ranges::range auto&& range) : small_vector() {
+        using Range = decltype(range);
+        if constexpr (std::ranges::sized_range<Range> || std::ranges::forward_range<Range>) {
+            auto n = std::ranges::distance(range);
+            reserve(n);
+            _end = std::ranges::uninitialized_copy(range, std::ranges::subrange(_end, _end + n)).out;
+        } else {
+            std::ranges::copy(range, std::back_inserter(*this));
+        }
+    }
```

Comment:
Hmm wait, can we potentially leak memory here, when the construction of an object from the range into the `small_vector` fails (throws an exception)?

`small_vector` claims that all of its member functions are strong-exception-safe. That claim is irrelevant here however, because this is not a member function, but a constructor. If the constructor fails (throws or propagates an exception), then there is no object for which to maintain consistency.

But precisely because there is no object yet, the `~small_vector()` destructor will not be invoked. Consequently, we won't call either (a) `clear() --> std::destroy(_begin, _end)`, or (b) `std::free(_begin)`.

(a) may or may not be a problem:

- [`std::ranges::uninitialized_copy`](https://en.cppreference.com/w/cpp/memory/ranges/uninitialized_copy) says that

> If an exception is thrown during the initialization, the objects already constructed are destroyed in an unspecified order.

- I note that [`std::ranges::copy`](https://en.cppreference.com/w/cpp/algorithm/ranges/copy) does not make a similar claim, however.

(b) seems like a problem to me. Assume that the input range is longer than the internal capacity of the `small_vector`. Then on both the `sized_range` / `forward_range` branch, and the other branch, we call `expand()` -- immediately, or eventually, respectively --, and allocate memory dynamically into `_begin`. If the element constructor throws after this point, then we leak `_begin`.

I think we need to catch all exceptions, free `_begin` if we don't use internal storage at that point, and rethrow the exception.

Reply:
When the constructor body is entered, the object is already constructed, since it delegates to the default constructor. So the destructor will be called.

Reply:
... in theory, on the single-pass branch, we could call `expand()` multiple times; if a call strictly after the first call throws `bad_alloc`, then the same leak is triggered, AFAICT. (IOW, it's not just the element constructor that can cause the leak.)

Reply:
I don't understand how we can leak.

Reply:
@avikivity

> When the constructor body is entered, the object is already constructed, since it delegates to the default constructor. So the destructor will be called.

You are right, my bad. I recognized that the constructor was a delegating one, but I missed the consequence for destruction / exceptions. It is explicitly covered in `n4950`, section `[except.ctor]`, paragraph 4:

> If the *compound-statement* of the *function-body* of a delegating constructor for an object exits via an exception, the object’s destructor is invoked. Such destruction is sequenced before entering a handler of the *function-try-block* of a delegating constructor for that object, if any.

Reply:
(I'm happy for this thread to be resolved, thanks for the education.)